include/mq/proto/Mailbox.h

```diff
@@ -0,0 +1,219 @@
+/*
+ * MacroQuest: The extension platform for EverQuest
+ * Copyright (C) 2002-2022 MacroQuest Authors
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License, version 2, as published by
+ * the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#pragma once
+
+#include "ProtoPipes.h"
+#include "Routing.pb.h"
+
+#include <string>
+#include <unordered_map>
+#include <queue>
+#include <memory>
+
+namespace mq {
+namespace mailbox {
+
+// we should assume that everything lives inside an Envelope here. All mail must be
+// in an envelope, no postcards (yet), but we open the Envelope to create ProtoMessages
+// when we put them on the queue
+class PostOffice
+{
+public:
+	using ReceiveCallback = std::function<void(const ProtoMessagePtr&)>;
+	using PostCallback = std::function<void(const std::string&)>;
+
+public:
+	class Mailbox
+	{
+		friend class PostOffice;
+
+	public:
+		Mailbox(
+			std::string_view localAddress,
+			ReceiveCallback receive,
+			PostCallback post
+		)
+			: m_localAddress(localAddress)
+			, m_receive(receive)
+			, m_post(post)
+		{}
+
+		template <typename ID, typename T>
+		void Post(const proto::Address& address, ID messageId, const T& obj)
+		{
+			m_post(Stuff(address, messageId, obj));
+		}
+
+		template <typename ID, typename T>
+		void PostReply(const ProtoMessagePtr& message, ID messageId, const T& obj, uint8_t status = 0)
+		{
+			if (auto returnAddress = message->GetReturn())
+			{
+				PostReply(message, *returnAddress, messageId, obj, status);
+			}
+			else
+			{
+				message->SendProtoReply(messageId, obj, status);
+			}
+		}
+
+		template <typename ID, typename T>
+		void PostReply(const PipeMessagePtr& message, const proto::Address& returnAddress, ID messageId, const T& obj, uint8_t status = 0)
```

Comment:
2 things that templating here gets us:
first, it allows us to set an ID type (that can be cast as uint32_t) that the individual post offices understand to ensure we aren't sending messages with invalid ID's, and to prevent us from needing to constantly cast to uint32_t at the call site.
second, it allows us to have actual serializable objects without needing to .SerializeAsString() at the call site.

This is a tiny amount of overhead in terms of multiple functions in the compiled code when you compare it against the convenience of the interface provided to the developer of new post offices, IMO.